go/store/nbs/journal_index_record.go

```diff
@@ -179,62 +182,177 @@ func validateIndexRecord(buf []byte) bool {
 	return crc(buf[:off]) == readUint32(buf[off:])
 }
 
-// processIndexRecords reads a sequence of index records from |r| and passes them to the callback. While reading records
-// it makes some basic assertions that the sequence is well-formed and indexes a contiguous region for the journal file.
-func processIndexRecords(ctx context.Context, r io.ReadSeeker, sz int64, cb func(o int64, r indexRec) error) (err error) {
-	var (
-		buf  []byte
-		off  int64
-		prev uint64
-	)
-
-	rdr := bufio.NewReader(r)
-	for off < sz {
-		// peek to read next record size
-		if buf, err = rdr.Peek(uint32Size); err != nil {
-			break
-		}
-
-		l := readUint32(buf)
-		if int64(l) > sz {
-			return fmt.Errorf("invalid record size %d for index file of size %d", l, sz)
-		}
-		if len(buf) < int(l) {
-			buf = make([]byte, l)
-		}
-		if _, err = io.ReadFull(rdr, buf); err != nil {
-			break
-		}
+type lookupMeta struct {
+	batchStart int
+	batchEnd   int
+	checkSum   uint32
+	latestHash hash.Hash
+}
 
-		// we do not zero-fill the journal index and expect
-		// only complete records that will checksum
-		if !validateIndexRecord(buf) {
-			return fmt.Errorf("failed to checksum index record at %d", off)
-		}
+const indexRecTypeSize = 1
+const (
+	indexRecChunk byte = iota
+	indexRecMeta
+)
 
-		var rec indexRec
-		if rec, err = readJournalIndexRecord(buf); err != nil {
+// processIndexRecords reads batches of chunk index lookups into the journal.
+// An index batch looks like |lookup|lookup|...|meta|. The first byte of a record
+// indicates whether it is a |lookup| or |meta|.
+func processIndexRecords(ctx context.Context, rd *bufio.Reader, sz int64, cb func(lookupMeta, []lookup, uint32) error) (err error) {
+	var batchCrc uint32
+	var batch []lookup
+	for {
+		recTag, err := rd.ReadByte()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
```

Comment:
We need to get the number of bytes we read after the last successfuly `indexRecMeta` callback back to the caller so that they can `file.Seek()` and `file.Truncate()` the output file to start at that point. Then we can start writing new records without causing a CRC failure in a later bootstrap.

Reply:
Same for the points where we return `nil` from `ErrUnexpectedEOF` down below.

Reply:
A bit of context that I was missing, the way this used to work was if anything went wrong we delete the index and exit process. The next startup loads all chunks into novel, and the next batch flush writes all chunks into a fresh index. Startup/clear/exit is the repeatable retry loop if anything goes wrong. Shitty for the next index flush, but it works.

The rewrite has different semantics that you pointed out. Now we are concerned with hanging index lookups after the last metadata record. I added a seek/truncate, which clears the handing lookups. We have to do an extra step where we add the hanging lookups back to the index for consistency. So I basically ignore most errors here, missing/malformed/io.EOF, we just truncate and rebuild. I think that should be equally repeatable, as long as there isn't a pathological loop where the index can't get a foothold for some reason.

Like one thing that is maybe a bit annoying in both versions is that we could rewrite the entire index, the server quits before writing a root value, and then next startup has to do it all over again. Reread all of the lookups, delete the index b/c no batch metadata, and then rebuilds the same index again.